lib/puppet/configurer.rb

```diff
@@ -55,13 +56,20 @@ def initialize(factory = Puppet::Configurer::DownloaderFactory.new)
   # Get the remote catalog, yo.  Returns nil if no catalog can be found.
   def retrieve_catalog(query_options)
     query_options ||= {}
-    # First try it with no cache, then with the cache.
-    unless (Puppet[:use_cached_catalog] and result = retrieve_catalog_from_cache(query_options)) or result = retrieve_new_catalog(query_options)
-      if ! Puppet[:usecacheonfailure]
-        Puppet.warning "Not using cache on failed catalog"
-        return nil
+    if (Puppet[:use_cached_catalog] && result = retrieve_catalog_from_cache(query_options))
```

Comment:
A bit of thinking aloud here, as I'm not 100% happy with what this ended up looking like.

There are a couple of cache-related situations that get resolved in this method. Once it has returned, we can't tell why we used the cache, so the logic of distinguishing between the following items must happen here:

1) `use_cached_catalog` is set *and* there is a catalog in the cache to be successfully retrieved.
  In this case, scenario 1 is true: We're using a cached catalog because it was explicitly requested.
2) `use_cached_catalog` is set *but* there is no cached catalog to be found.
  This is the section scenario, where we *asked* for a cached catalog but will instead have to get a new one.
3) We didn't ask for a cached catalog, and successfully downloaded a new one.
  In this case, no cached catalog is involved at all.
4) We didn't ask for a cached catalog, but remote compilation failed so we fell back using `usecacheonfailure`
  In this case, we're using the cache as a fallback.


Later on in the caller of this method, cases 1 and 4 will become indistinguishable. There is also the edge case of when we need a cached catalog (it was specifically requested or the new catalog failed to compile) but there is no cached catalog available. In this case we need to specify that no cached catalog was used.

Although the keys ('not used', 'used because...') will be condensed into one word values, I'm not sure I like hardcoding them in here like this. Should this logic live elsewhere, or be abstracted into a conditional below when editing the report?